flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceStreamTask.java

```diff
@@ -121,10 +123,12 @@ protected void processInput(MailboxDefaultAction.Controller controller) throws E
 		sourceThread.setTaskDescription(getName());
 		sourceThread.start();
 		sourceThread.getCompletionFuture().whenComplete((Void ignore, Throwable sourceThreadThrowable) -> {
-			if (sourceThreadThrowable == null || isFinished) {
-				mailboxProcessor.allActionsCompleted();
-			} else {
+			if (isCanceled() && ExceptionUtils.findThrowable(sourceThreadThrowable, InterruptedException.class).isPresent()) {
+				mailboxProcessor.reportThrowable(new CancelTaskException(sourceThreadThrowable));
```

Comment:
We probably need to restore interruption flag `Thread.interrupt()`.

Reply:
Interruption is internal to `SourceStreamTask`, so I think the flag shouldn't be set.